File: Postgres/ServerManager.m
//
//  ServerManager.m
//  Postgres
//
//  Created by Chris on 23.05.16.
//
//

#import "ServerManager.h"
#import "PostgresServer.h"


@implementation ServerManager

#pragma mark - Singleton class methods

+ (ServerManager *)sharedManager {
	static ServerManager *sharedManager = nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		sharedManager = [[self.class hiddenAlloc] init];
	});
	return sharedManager;
}


+ (id)alloc {
	NSLog(@"%@ is a singleton - use +sharedInstance", self.class.description);
	return nil;
}


+ (id)hiddenAlloc {
	return [super alloc];
}



#pragma mark - Instance methods

- (id)init {
	self = [super init];
	if (self) {
		self.servers = [[NSMutableArray alloc] init];
	}
	return self;
}


- (void)refreshStatus {
	for (PostgresServer *server in self.servers) {
		[server serverStatus];
	}
}


- (void)startServers {
	for (PostgresServer *server in self.servers) {
		if (server.runAtStartup) {
			[server startWithCompletionHandler:^(BOOL success, NSError *error) {
				NSLog(@"%@ started with success=%d error=%@", server.name, success, error);
			}];
		}
	}
}


- (void)stopServers {
	for (PostgresServer *server in self.servers) {
		if (server.stopAtQuit) {
			[server stopWithCompletionHandler:^(BOOL success, NSError *error) {}];
		}
	}
}


- (void)saveServers {
	NSData *data = [NSKeyedArchiver archivedDataWithRootObject:self.servers];
	[[NSUserDefaults standardUserDefaults] setObject:data forKey:@"servers"];
}


- (void)loadServers {
	NSData *data = [[NSUserDefaults standardUserDefaults] objectForKey:@"servers"];
	NSMutableArray *arr = [NSKeyedUnarchiver unarchiveObjectWithData:data];
	
	if ([arr count] > 0) {
		[self.servers addObjectsFromArray:arr];
	}
}


- (void)loadServersForHelperApp {
	NSUserDefaults *defaults = [[NSUserDefaults alloc] initWithSuiteName:@"com.postgresapp.Postgres"];
	NSData *data = [defaults objectForKey:@"servers"];
	NSMutableArray *arr = [NSKeyedUnarchiver unarchiveObjectWithData:data];
	
	if ([arr count] > 0) {
		[self.servers addObjectsFromArray:arr];
	}
}

@end

File: Postgres/ServerManager.swift
//
//  ServerManager.swift
//  Postgres
//
//  Created by Chris on 23/06/16.
//  Copyright © 2016 postgresapp. All rights reserved.
//

import Cocoa

class ServerManager: NSObject {
	
	static let shared = ServerManager()
	
	@objc dynamic var servers: [Server] = []
	
	
	func refreshServerStatuses() {
		for server in servers {
			server.updateServerStatus()
		}
	}
	
	func saveServers() {
		var plists: [[AnyHashable: Any]] = []
		for server in servers {
			plists.append(server.asPropertyList)
		}
		UserDefaults.standard.set(plists, forKey: "Servers")
	}
	
	
	func loadServers() {
		servers.removeAll()
		guard let plists = UserDefaults.shared.array(forKey: "Servers") as? [[AnyHashable: Any]] else {
			NSLog("PostgresApp could not load servers from user defaults.")
			return
		}
		for plist in plists {
			guard let server = Server(propertyList: plist) else {
				NSLog("PostgresApp could not load server from user defaults.")
				continue
			}
			servers.append(server)
		}
	}
}

